src/adagrams.js

```diff
@@ -1,15 +1,177 @@
+    drawLetter = flatLetterPool[indexToSelect];
+    resLetters.push(drawLetter);
+    flatLetterPool.splice(indexToSelect, 1);
+  }
+  return resLetters;
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
   // Implement this method for wave 2
+
+  let upperInput;
+  upperInput = input.toUpperCase();
+
+  // As we loop thru char in upperInput, delete each matching char in array of resLetters;
+  // Return False if unable to find, else True;
+  for (let char of upperInput) {
+    if (!lettersInHand.includes(char)) {
+      return false;
+    } else if (lettersInHand.includes(char)) {
+      lettersInHand.splice(lettersInHand.indexOf(char), 1);
+    }
+  }
+  return true;
 };
 
 export const scoreWord = (word) => {
   // Implement this method for wave 3
+  let sumScore = 0;
+  let upperWord;
+
+  if (word !== undefined) {
+    upperWord = word.toUpperCase();
+    if (upperWord.length === 0) {
+      sumScore = 0;
+    } else if (upperWord.length >= 7 && upperWord.length <= 10) {
+      sumScore += 8;
+    }
+    for (let char of upperWord) {
+      if (char in scorePool) {
+        sumScore += scorePool[char];
+      }
+    }
+    return sumScore;
+  }
 };
 
 export const highestScoreFrom = (words) => {
   // Implement this method for wave 4
+  // words is array of word; return should be {word: name; score: num}
+  // loop thru each word to calculate score, save in scoreDict {score: [array of words]}
+  // find highest score / scores
+  // if highest score array length = 1, return as result;
+  // else, going into tie logic;
+
+  // tie logic: <loop thru highest score array>
+  // 1. check if there is word with 10 letters -> return immediately;
+  // 2. fewer letter: mark fewer word name and length, and update along loop; -> return the fewest;
+  // 3. index first: above logic should satisfied index;
+
+  let scoreDict = {};
+  let wordScore;
+  let maxScore;
+  let maxScoreArray;
+  let resDict = { word: "", score: 0 };
+  let fewerWordLength = 10;
+  let fewerWordName;
+
+  for (let word of words) {
+    wordScore = scoreWord(word);
+    if (!(wordScore in scoreDict)) {
+      scoreDict[wordScore] = [];
+    }
+    scoreDict[wordScore].push(word);
+  }
+  // array spread operator
+  maxScore = Math.max(...Object.keys(scoreDict));
+  maxScoreArray = scoreDict[maxScore];
+  if (maxScoreArray.length === 1) {
+    resDict["word"] = maxScoreArray.toString();
+    resDict["score"] = maxScore;
+    return resDict;
+  } else {
+    //tie logic thru maxScoreArray
+    for (let maxWord of maxScoreArray) {
+      if (maxWord.length === 10) {
+        resDict["word"] = maxWord;
+        resDict["score"] = maxScore;
+        return resDict;
+      } else {
+        //length != 10
+        if (maxWord.length < fewerWordLength) {
+          fewerWordLength = maxWord.length;
+          fewerWordName = maxWord;
+          resDict["word"] = fewerWordName;
+          resDict["score"] = scoreWord(fewerWordName);
```

Comment:
Very interesting solution!  Slight possible optimization - it's not necessary to call `scoreWord` here, because the score is guaranteed to be `maxScore`.